### PR TITLE
[simple-cache-provider] New implementation

### PR DESCRIPTION
**Pull request by acdlite**
## Based on #13293

An updated implementation of simple-cache-provider (final name TBD) using `Context.set` and `Context.read`.

- [x] Subscription support. `load` accepts a promise or an observable as its return value.
- [x] Default to LRU cache that is shared across resources
- [x] Use `calculateChangedBits` API to optimize updates

Things I'll probably defer until later since they aren't required for an MVP:
- Support overriding global cache implementation
- Support overriding cache implementation for a subtree

### CODE REVIEW

**Comment by react-linter** [comment]
React: size: :small_red_triangle:+7.3%, gzip: :small_red_triangle:+5.7%

ReactDOM: size: :small_red_triangle:+1.1%, gzip: :small_red_triangle:+0.9%


  <details>
  <summary>Details of bundled changes.</summary>

  <p>Comparing: 2a2ef7e0fd86612404676b15bb3dd9df7ade536e...f6d6304afc26ef80f1673845143aecc4a88c28fa</p>


  
## react
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react.development.js | +3.2% | +3.3% | 57.81 KB | 59.65 KB | 16.16 KB | 16.7 KB | UMD_DEV
**react.production.min.js** | **:small_red_triangle:+7.3%** | **:small_red_triangle:+5.7%** | **7.01 KB** | **7.52 KB** | **2.96 KB** | **3.13 KB** | **UMD_PROD**
react.development.js | +3.5% | +3.5% | 52 KB | 53.84 KB | 14.34 KB | 14.84 KB | NODE_DEV
**react.production.min.js** | **:small_red_triangle:+8.4%** | **:small_red_triangle:+6.9%** | **6.02 KB** | **6.52 KB** | **2.57 KB** | **2.74 KB** | **NODE_PROD**
React-dev.js | +3.7% | +3.7% | 49.82 KB | 51.66 KB | 13.75 KB | 14.26 KB | FB_WWW_DEV
**React-prod.js** | **:small_red_triangle:+10.3%** | **:small_red_triangle:+8.5%** | **13.69 KB** | **15.1 KB** | **3.8 KB** | **4.12 KB** | **FB_WWW_PROD**

## react-dom
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-dom.development.js | +0.6% | +0.5% | 640.7 KB | 644.42 KB | 150.38 KB | 151.15 KB | UMD_DEV
**react-dom.production.min.js** | **:small_red_triangle:+1.1%** | **:small_red_triangle:+0.9%** | **95.87 KB** | **96.93 KB** | **31.08 KB** | **31.37 KB** | **UMD_PROD**
react-dom.development.js | +0.6% | +0.5% | 636.84 KB | 640.56 KB | 149.2 KB | 149.98 KB | NODE_DEV
**react-dom.production.min.js** | **:small_red_triangle:+1.1%** | **:small_red_triangle:+1.2%** | **95.85 KB** | **96.93 KB** | **30.58 KB** | **30.93 KB** | **NODE_PROD**
ReactDOM-dev.js | +0.6% | +0.6% | 644 KB | 647.79 KB | 147.54 KB | 148.38 KB | FB_WWW_DEV
**ReactDOM-prod.js** | **:small_red_triangle:+1.0%** | **:small_red_triangle:+1.2%** | **277.59 KB** | **280.42 KB** | **52.08 KB** | **52.7 KB** | **FB_WWW_PROD**
react-dom.profiling.min.js | +1.1% | +1.1% | 96.99 KB | 98.06 KB | 30.99 KB | 31.32 KB | NODE_PROFILING
ReactDOM-profiling.js | +1.0% | +1.2% | 280.88 KB | 283.71 KB | 52.8 KB | 53.43 KB | FB_WWW_PROFILING

## react-art
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-art.development.js | +0.9% | +0.8% | 427.11 KB | 430.83 KB | 96.47 KB | 97.26 KB | UMD_DEV
**react-art.production.min.js** | **:small_red_triangle:+1.3%** | **:small_red_triangle:+1.2%** | **83.25 KB** | **84.36 KB** | **25.75 KB** | **26.05 KB** | **UMD_PROD**
react-art.development.js | +1.0% | +1.0% | 359.63 KB | 363.35 KB | 79.41 KB | 80.19 KB | NODE_DEV
**react-art.production.min.js** | **:small_red_triangle:+2.3%** | **:small_red_triangle:+2.2%** | **48.23 KB** | **49.35 KB** | **15.15 KB** | **15.47 KB** | **NODE_PROD**
ReactART-dev.js | +1.1% | +1.1% | 349.71 KB | 353.5 KB | 74.25 KB | 75.07 KB | FB_WWW_DEV
**ReactART-prod.js** | **:small_red_triangle:+2.1%** | **:small_red_triangle:+2.3%** | **148.93 KB** | **152 KB** | **25.42 KB** | **26.01 KB** | **FB_WWW_PROD**

## react-test-renderer
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-test-renderer.development.js | +1.0% | +0.9% | 362.85 KB | 366.57 KB | 79.68 KB | 80.43 KB | UMD_DEV
**react-test-renderer.production.min.js** | **:small_red_triangle:+2.3%** | **:small_red_triangle:+2.4%** | **48.76 KB** | **49.88 KB** | **15.05 KB** | **15.41 KB** | **UMD_PROD**
react-test-renderer.development.js | +1.0% | +1.0% | 358.98 KB | 362.7 KB | 78.7 KB | 79.46 KB | NODE_DEV
**react-test-renderer.production.min.js** | **:small_red_triangle:+2.3%** | **:small_red_triangle:+2.5%** | **48.47 KB** | **49.59 KB** | **14.9 KB** | **15.27 KB** | **NODE_PROD**
ReactTestRenderer-dev.js | +1.0% | +1.0% | 363.98 KB | 367.76 KB | 77.63 KB | 78.42 KB | FB_WWW_DEV

## react-reconciler
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
react-reconciler.development.js | +1.1% | +1.1% | 347.97 KB | 351.69 KB | 75.3 KB | 76.14 KB | NODE_DEV
**react-reconciler.production.min.js** | **:small_red_triangle:+2.5%** | **:small_red_triangle:+2.4%** | **47.44 KB** | **48.65 KB** | **14.37 KB** | **14.71 KB** | **NODE_PROD**
react-reconciler-persistent.development.js | +1.1% | +1.1% | 346.59 KB | 350.3 KB | 74.75 KB | 75.56 KB | NODE_DEV
**react-reconciler-persistent.production.min.js** | **:small_red_triangle:+2.5%** | **:small_red_triangle:+2.4%** | **47.45 KB** | **48.66 KB** | **14.37 KB** | **14.71 KB** | **NODE_PROD**

## simple-cache-provider
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
simple-cache-provider.development.js | +78.9% | +61.5% | 8.79 KB | 15.73 KB | 2.83 KB | 4.57 KB | NODE_DEV
**simple-cache-provider.production.min.js** | **:small_red_triangle:+79.1%** | **:small_red_triangle:+53.1%** | **1.63 KB** | **2.92 KB** | **827 B** | **1.24 KB** | **NODE_PROD**
SimpleCacheProvider-dev.js | +86.0% | +72.9% | 7.92 KB | 14.73 KB | 2.4 KB | 4.14 KB | FB_WWW_DEV
**SimpleCacheProvider-prod.js** | **:small_red_triangle:+120.6%** | **:small_red_triangle:+73.0%** | **3.65 KB** | **8.04 KB** | **1.11 KB** | **1.92 KB** | **FB_WWW_PROD**

## react-native-renderer
File | Filesize Diff | Gzip Diff | Prev Size | Current Size | Prev Gzip | Current Gzip | ENV
 ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  ---  |  --- 
ReactNativeRenderer-dev.js | +0.8% | +0.8% | 482.14 KB | 485.92 KB | 106.48 KB | 107.3 KB | RN_FB_DEV
**ReactNativeRenderer-prod.js** | **:small_red_triangle:+1.3%** | **:small_red_triangle:+1.5%** | **213.76 KB** | **216.5 KB** | **37.44 KB** | **38.01 KB** | **RN_FB_PROD**
ReactNativeRenderer-dev.js | +0.8% | +0.8% | 481.88 KB | 485.66 KB | 106.43 KB | 107.24 KB | RN_OSS_DEV
**ReactNativeRenderer-prod.js** | **:small_red_triangle:+1.3%** | **:small_red_triangle:+1.6%** | **203.8 KB** | **206.55 KB** | **35.82 KB** | **36.41 KB** | **RN_OSS_PROD**
ReactFabric-dev.js | +0.8% | +0.8% | 472.36 KB | 476.14 KB | 104.05 KB | 104.87 KB | RN_FB_DEV
**ReactFabric-prod.js** | **:small_red_triangle:+1.4%** | **:small_red_triangle:+1.7%** | **194.92 KB** | **197.68 KB** | **34.27 KB** | **34.87 KB** | **RN_FB_PROD**
ReactFabric-dev.js | +0.8% | +0.8% | 472.39 KB | 476.17 KB | 104.07 KB | 104.88 KB | RN_OSS_DEV
**ReactFabric-prod.js** | **:small_red_triangle:+1.4%** | **:small_red_triangle:+1.7%** | **194.95 KB** | **197.71 KB** | **34.29 KB** | **34.88 KB** | **RN_OSS_PROD**
ReactNativeRenderer-profiling.js | +1.3% | +1.6% | 207.31 KB | 210.06 KB | 36.51 KB | 37.1 KB | RN_OSS_PROFILING
ReactFabric-profiling.js | +1.4% | +1.9% | 197.97 KB | 200.73 KB | 34.91 KB | 35.56 KB | RN_OSS_PROFILING
ReactNativeRenderer-profiling.js | +1.3% | +1.5% | 217.23 KB | 219.97 KB | 38.13 KB | 38.71 KB | RN_FB_PROFILING
ReactFabric-profiling.js | +1.4% | +1.9% | 197.93 KB | 200.69 KB | 34.89 KB | 35.54 KB | RN_FB_PROFILING

  </details>
  
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by plievone** [comment]
This is very exciting work, enabling streaming stores with automatic dependency tracking, granular fetching, still keeping the UI consistent at all times, etc. Sorry if this is too early to ask, but I'm interested in how many contexts will be feasible with this design, are a few hundred contexts/stores ok performance-wise or too much for this design?

File: packages/react/src/ReactRootList.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * @flow
+ */
+
+import type {ReactContext} from 'shared/ReactTypes';
+
+type GlobalRoot = {
+  isMounted: boolean,
+  setContext<T>(
+    context: ReactContext<T>,
+    oldVvalue: T,
```
Comment (plievone) [inline packages/react/src/ReactRootList.js]: You might have a typo here, `oldVvalue` (and a type problem if not seen by flow)

**Comment by plievone** [comment]
@acdlite, I would like to start experimenting with simple-cache-provider. Would it make sense for me to rebase this implementation of yours on master or just use the existing version? This one has subscriptions etc.

**Comment by jaredpalmer** [comment]
Just summarizing these API changes for reference as I wasn't able to find them documented elsewhere:

- `createResource` now has support for subscriptions. You return an object with a `subscribe` key which sets up the subscription, and it can return another fn for teardown.
- Breaking Change: `Resource.read()` no longer needs `cache` instance as first argument, just the input.

**New Resource API**

- `Resource.preload(input)`: Preload a value in the cache
- `Resource.read(input)`: Read a value from the cache

**New Cache API**

- `useCache()`: Get cache via context (calls `Context.unstable_read()`)
- `cache.set(resource, input, value)`: Set a value in the cache
- `cache.purge()`: Wipe the LRU cache
- `cache.invalidate(resource, input)`: Invalidate an item in the cache, given a resource and input (will run the hash fn, if it was provided to resource) 
- `setGlobalCacheLimit(limit)`: Set global LRU cache size

**Comment by acdlite** [comment]
@plievone This branch is the closest to the final version

**Comment by jaredpalmer** [comment]
@acdlite Why not add `Resource.invalidate(input)` and `Resource.set(input, value)` to the Resource API? For basic CRUD stuff, that would avoid having to use the cache directly. It might make the mental model a bit easier too...one less thing to pass around in your application.

```jsx
function Thing(props) {
 const data = ThingResource.read(props);
 function update() {
   PatchUpdateThing().then(() => {
     // Trigger a reload
     ThingResource.invalidate(props)
   })
 }
 return <Something data={data} refresh={update} />
}
```

**Comment by acdlite** [comment]
@jaredpalmer Short answer: the cache is contextual, not global. Longer answer: in practice the cache you read from context is almost always the global one. But if you start with a global API, then later want to render your component in a contextual setting, now you have to refactor all your components to the contextual form. Requiring the contextual version always adds a little bit of typing overhead but it avoids the refactor hazard. (Similar reasoning to why Redux components use `connect()` instead of reading from the store directly.)

**Comment by plievone** [comment]
@acdlite Here is a rebased version, feel free to pull into your workflow: https://github.com/plievone/react/tree/rewrite-scp-rebase

I aimed to keep your whitespace, commented code, etc. The two commits are there, as you have different pull requests with review going on for those.

`yarn test` passes except one related to react-profiler tracing:
```
Profiler › interaction tracing › suspense › traces both the temporary placeholder and the finished render for an interaction
Invariant Violation: Context.unstable_read(): Context can only be read while React is rendering, e.g. inside the render method or getDerivedStateFromProps.
```

`yarn lint` passes except one error related to flow typings (`CacheImplementation`  is defined but never used)

`yarn flow` has errors that have been there I guess.

I also updated the suspense fixture, and had to fix one trace call missing performance.now() to test cache purging. But seemed to work now in quick testing.

So there's some work to do to fix typings, failing test, addressing comments (renaming things, reimplementing things perhaps), but hopefully this is a quicker starting point now.

**Comment by jaredpalmer** [comment]
In my React Conf talk, would it be better to show off code from this PR or what's currently in master?

**Comment by acdlite** [comment]
@jaredpalmer Uh, you and I need to chat about your React Conf talk at some point so we can coordinate :D

**Comment by jaredpalmer** [comment]
Will DM.

**Comment by plievone** [comment]
@acdlite I wanted to experiment with hooks too, so here's the latest rebased version of your PR: https://github.com/plievone/react/tree/rewrite-scp-rebase2

I aimed to keep your whitespace, commented code, etc, but changed some lines according to changes in master in the meantime. The two commits are there, as you have different pull requests with review going on for those.

yarn test passes. When rebasing, I converted the added react-cache tests to use react-test-renderer to pass, except a couple of tests that I couldn't make pass with react-test-renderer so kept them in a separate file with react-noop-renderer.

yarn flow has errors, haven't touched those.

I have been enjoying experimenting with this API (especially automatic updates when changing/invalidating cache keys). To perhaps continue from this rebased code (for example the converted tests might help you), would you like me to create pull request somewhere?

**Comment by zaguiini** [comment]
What's the status of this PR?

**Comment by acdlite** [comment]
Will likely ship some version of this package as part of initial Concurrent Mode release, but not this one in this PR. Closing for now.

**Pull request closed without merging**